### ISSUE TITLE
test/scylla-gdb/run should clearly complain when debug symbols are missing

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#10863]
The scylla-gdb test suite only works when the Scylla executable is built with symbols - so it works with release and debug build modes, but not with dev mode. On a dev build, each and every one of the test functions fails with an error like:

```
    class intrusive_list:
>       size_t = gdb.lookup_type('size_t')
```

It is annoying to have dozens of individual tests fail with a weird error, so **test/scylla-gdb/run** has special code to check the debug symbols before it runs any test:

```python
if not 'debug_info' in subprocess.run(['file', run.scylla],
        capture_output=True, text=True).stdout:
    print(f'Scylla executable was compiled without debugging information (-g)')
    print(f'so cannot be used to test gdb. Please set SCYLLA environment variable.')
```

But for some reason, this check no longer works... Today, file's output contains "debug_info" even in dev build:

>  $ file build/dev/scylla
> build/dev/scylla: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, BuildID[sha1]=e66bf830fe916b7094cad86a504ed1bb063d4bca, with **debug_info**, not stripped, too many notes (256)

**Comment by nyh** [comment]
When linking code that parts of it have debug symbols, and other parts don't, the result shows it has "debug_info" even though the debug information is very partial. `readelf -wl build/dev/scylla` says that we got some debug info from Rust (CC @wmitros):
```
Raw dump of debug contents of section .debug_line:
...
  1     /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/adapters
  2     /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/str
  3     /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/slice
...
```

We have two options now:
1. Find a way to compile rust without debug symbols, so as not to add these debug symbols to the dev build.
2. Find a different way to check if Scylla was built with debug information of the real symbols we care about. This different way should be relatively fast (not to start gdb just to check, which is annoyingly slow).

**Comment by nyh** [comment]
It appears that we get the debug symbols from /usr/lib64/libwasmtime.a with which we always link.

**Comment by nyh** [comment]
@wmitros , @psarna ,

I just realized that /usr/lib64/libwasmtime.a isn't actually a library from my system - it was installed by install-dependencies.sh. Install-dependencies.sh can also strip this library from the unwanted debugging symbols (which libraries normally do not have).

The problem is that even if I change install-dependencies.sh, what if developers never run it again? install-dependencies.sh isn't something you usually need to run periodically. Which begs a question - why is the fetching of wasmtime part of install-dependencies.sh and not part of build.ninja? Why do we install this non-standard library in /usr/lib64?

**Comment by psarna** [comment]
> @wmitros , @psarna ,
> 
> I just realized that /usr/lib64/libwasmtime.a isn't actually a library from my system - it was installed by install-dependencies.sh. Install-dependencies.sh can also strip this library from the unwanted debugging symbols (which libraries normally do not have).

Indeed, we could have done that, I just didn't take debug symbol into account at all when setting it up.

> 
> The problem is that even if I change install-dependencies.sh, what if developers never run it again?

They will, once we add a new dependency that breaks compilation. But if that never happens, well, they would simply have an unstripped version of libwasmtime.a in their system.

> install-dependencies.sh isn't something you usually need to run periodically.

Well, not periodically as in "every first Friday of a month", but developers are generally expected to run it every time it changes. Usually it comes naturally, because compilations start failing due to missing dependencies, but this time it would simply be the matter of having a dependency with unwanted debug symbols, which isn't a major issue (at least to me, I never noticed).

> Which begs a question - why is the fetching of wasmtime part of install-dependencies.sh and not part of build.ninja? Why do we install this non-standard library in /usr/lib64?

libwasmtime is a library we depend on, so ideally it should just be added as a dependency on a package available on Linux distributions we support. Since it is not available (or at least it wasn't), I figured that the closest approximation of installing a package would be downloading given binary release of this library from source, and installing it manually, by putting it in one of the /lib directories.

But, having said all that, let me also clarify that this dependency of ours is transient, and we added it only to allow early experimental Wasm support for user-defined functions. We are in the process of dropping the dependency on libwasmtime bindings in favor of using the original Rust library (https://crates.io/crates/wasmtime) and creating our own bindings, with async support, in order to be able to avoid reactor stalls. The code is still work in heavy progress, you can take a peek at the code here:
Rust bindings: https://github.com/wmitros/scylla/blob/rust-wasmtime/rust/wasmtime_bindings/src/lib.rs
Example of using the bindings in Seastar: https://github.com/wmitros/scylla/blob/d9a676be0f0a752ebd9037da4385b6f00d0e3eb4/lang/wasm.cc#L236-L249
Bottom line is that once we have all these changes merged, we won't depend on wasmtime anymore and we could purge the whole root cause of this issue out of install-dependencies.sh.

@wmitros by the way, we need to make sure that once we compile the Rust bindings ourselves, we also drop debug symbols from the compiled objects in dev mode, otherwise we'd still have a similar issue.

**Comment by wmitros** [comment]
> @wmitros by the way, we need to make sure that once we compile the Rust bindings ourselves, we also drop debug symbols from the compiled objects in dev mode, otherwise we'd still have a similar issue.

In the same patch that you've mentioned I'm adding specific options for different build modes (currently we're building with the same options in every mode). One of them is debug symbol stripping so that should help

> When linking code that parts of it have debug symbols, and other parts don't, the result shows it has "debug_info" even though the debug information is very partial.
> 
> `readelf -wl build/dev/scylla` says that we got some debug info from Rust (CC @wmitros):

It's a bit surprising to me that there were symbols in rust, because in the configuration of rust build we're using an option `debug=false`, but as I mentioned above, additional symbol stripping should take care of that

**Comment by nyh** [comment]
> @wmitros , @psarna ,
> 
> I just realized that /usr/lib64/libwasmtime.a isn't actually a library from my system - it was installed by install-dependencies.sh. Install-dependencies.sh can also strip this library from the unwanted debugging symbols (which libraries normally do not have).

I just realized something even stranger - we don't immediately install it in /usr/lib64 - we actually put it in /opt/scylladb/dependencies and then copy it to /usr/lib64. We don't use what we leave in /opt/scylladb/dependencies, and even leave the compressed tar there as well - and potentially several versions there if we ever change the version we pick up. This doesn't look a good idea at all, but of course not directly related to this patch :-( Why don't we leave the library - and only the library - in /opt/scylladb/depedencies? Anything else we need temporarily (like the tar) we can download to a temporary directory.

**Comment by avikivity** [comment]
No production impact, not backporting.
